### PR TITLE
feat: update export controller

### PR DESCRIPTION
**Pull request by zhiying-lin**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
added the util func to support atm feature

**Requirements**:

- [ ] run `make reviewable` for basic local test
- [ ] Read and followed fleet-networking's [Code of conduct](https://github.com/Azure/fleet-networking/blob/main/CODE_OF_CONDUCT.md).
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

### Special notes for your reviewer

The new func will be called in the next PR.

### CODE REVIEW

File: pkg/controllers/member/serviceexport/controller.go [hunk 1]
```diff
@@ -42,6 +48,8 @@ const (
 
 	// ControllerName is the name of the Reconciler.
 	ControllerName = "serviceexport-controller"
+
+	publicLoadBalancerName = "kubernetes"
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:52]: Do we only support AKS cluster for now or any clusters held in Azure? Another thing is looks like multi-SLB work is undergoing, so we will need to dynamically decide LB name from service. For now, it's ok.

Multi-SLB info will be included in azure cloud config too: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/provider/azure.go#L295

File: pkg/controllers/member/serviceexport/controller.go [hunk 2]
```diff
@@ -256,6 +268,70 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	if export.Spec.IsInternalLoadBalancer {
+		// no need to populate the PublicIPResourceID and IsDNSLabelConfigured which are only applicable for external load balancer
+		return nil
+	}
+	id, err := r.fetchPublicIPResourceID(ctx, service)
+	if err != nil {
+		return err
+	}
+	if id == nil || len(*id) == 0 {
+		return fmt.Errorf("public ip resource id %v has not been configured in the load balancer yet", id)
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:286]: Does it mean cloud provider is still working on it? Do we retry? Do we check if service already has lbIP in status already?

File: pkg/controllers/member/serviceexport/controller.go [hunk 3]
```diff
@@ -256,6 +268,70 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+}
+
+func (r *Reconciler) fetchPublicIPResourceID(ctx context.Context, service *corev1.Service) (*string, error) {
+	serviceKObj := klog.KObj(service)
+	lb, err := r.AzureLoadBalancerClient.Get(ctx, r.LoadBalancerResourceGroupName, publicLoadBalancerName, nil)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Azure load balancer", "service", serviceKObj)
+		return nil, err
+	}
+	frontedIPName := "a" + string(service.UID)
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:305]: the dash "-" should be removed: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/6c002772ad81dc56e74a3a726a3b831387989560/pkg/provider/azure_loadbalancer.go#L402, https://github.com/kubernetes/kubernetes/blob/15d08bf7c8813b0533dc147a03d9f42aae735ecd/staging/src/k8s.io/cloud-provider/cloud.go#L89

File: pkg/controllers/member/serviceexport/controller.go [hunk 4]
```diff
@@ -256,6 +268,70 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	if lb.Properties != nil {
+		for _, frontendIPConfig := range lb.Properties.FrontendIPConfigurations {
+			if frontendIPConfig.Name != nil && *frontendIPConfig.Name == frontedIPName {
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:308]: hmmmm, sorry but I realized one thing: services many share frontends. Cx can create multiple services just specifying the same public IP. In this case, these services will share the same frontend named after the first service obj UID. See https://github.com/kubernetes-sigs/cloud-provider-azure/blob/6c002772ad81dc56e74a3a726a3b831387989560/pkg/provider/azure_loadbalancer.go#L4113 

To resolve this case, we need to list and find the PIP directly. Do you think if we keep this for now given that sharing frontend is a less common usage or we directly turn to list PIP and find the PIP resource according to the public IP addr in the svc status? In this case, we can get rid of the loadbalancer at all.

File: pkg/controllers/member/serviceexport/controller.go [hunk 5]
```diff
@@ -256,6 +264,65 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		// Assuming once the service status is updated, the controller will be triggered again.
+		klog.V(2).InfoS("The load balancer IP is not assigned yet", "service", serviceKObj)
+		return nil
+	}
+
+	if service.Status.LoadBalancer.Ingress[0].IP == "" {
+		err := errors.New("the service ingress is not nil but with empty IP")
+		klog.ErrorS(controller.NewUnexpectedBehaviorError(err), "Failed to get the load balancer IP from service", "service", serviceKObj, "status", service.Status)
+		return nil
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:288]: Do we want to return err here?

File: pkg/controllers/member/serviceexport/controller.go [hunk 6]
```diff
@@ -256,6 +264,65 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+		klog.V(2).InfoS("The public IP is in the progressing", "service", serviceKObj, "ip", service.Status.LoadBalancer.Ingress[0].IP)
+		// Assuming once the service status is updated, the controller will be triggered again in instead of retrying here
+		// to avoid sending Azure requests.
+		return nil
+	}
+	export.Spec.PublicIPResourceID = pip.ID
+	// Note the user can set the dns label via the Azure portal or Azure CLI without updating service.
+	// This information may be stale as we don't monitor the public IP address resource.
+	export.Spec.IsDNSLabelConfigured = pip.Properties != nil && pip.Properties.DNSSettings != nil && pip.Properties.DNSSettings.DomainNameLabel != nil
+	return nil
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:305]: DNS label is required right? In that case, we should return err and retry?

File: pkg/controllers/member/serviceexport/controller.go [hunk 7]
```diff
@@ -256,6 +264,65 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+}
+
+// TODO: can improve the performance by caching the public IP address resource ID.
+func (r *Reconciler) lookupPublicIPResourceIDByLoadBalancerIP(ctx context.Context, service *corev1.Service) (*armnetwork.PublicIPAddress, error) {
+	serviceKObj := klog.KObj(service)
+	pips, err := r.AzurePublicIPAddressClient.List(ctx, r.ResourceGroupName)
```
Comment (jwtty) [inline pkg/controllers/member/serviceexport/controller.go:311]: cx may also set pip rg in the service annotation: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/provider/azure_loadbalancer.go#L3492

File: pkg/controllers/member/serviceexport/controller.go (hunk 5 above)
Comment (zhiying-lin) [reply pkg/controllers/member/serviceexport/controller.go:288]: Assume it's not possible in production, so logged as unexpected error.

If the IP of service is updated, the controller will be retriggered.

File: pkg/controllers/member/serviceexport/controller.go (hunk 6 above)
Comment (zhiying-lin) [reply pkg/controllers/member/serviceexport/controller.go:305]: serviceExport can be used by the existing MCS. The DNS label is required by the ATM, but not MCS.

**Pull request merged**